### PR TITLE
Extend profiler toolbar item

### PR DESCRIPTION
**Pull request by chr-hertel**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | -
| Documentation   | -
| License         | MIT

**Depends on [stopwatch plugin PR](https://github.com/php-http/stopwatch-plugin/pull/3)**

#### What's in this PR?

Extended profiler toolbar item to display more details and duration

bildschirmfoto 2017-03-17 um 17 47 31

### CODE REVIEW

File: Collector/StackPlugin.php [hunk 1]
```diff
@@ -49,12 +49,22 @@ public function __construct(Collector $collector, Formatter $formatter, $client)
...
+            $this->formatter->formatRequest($request)
+        );
 
         $this->collector->addStack($stack);
 
         return $next($request)->then(function (ResponseInterface $response) use ($stack) {
             $stack->setResponse($this->formatter->formatResponse($response));
+            $stack->setResponseCode($response->getStatusCode());
+
+            if ($response->hasHeader('X-Duration')) {
```
Comment (dbu) [inline Collector/StackPlugin.php]: where does the x-duration header come from?

Comment (dbu) [inline Collector/StackPlugin.php]: ah, now i got it, this depends on https://github.com/php-http/stopwatch-plugin/pull/3

**Comment by dbu** [comment]
@Nyholm it depends on https://github.com/php-http/stopwatch-plugin/pull/3 and we currently are trying to figure out how to track requests and responses in the plugin chain...

**Comment by fbourigault** [comment]
What do you thing about removing the stopwatch plugin from the profiled stacks and directly use the stopwatch component in the `StackPlugin`?

```php
    /**
     * {@inheritdoc}
     */
    public function handleRequest(RequestInterface $request, callable $next, callable $first)
    {
        $stack = new Stack(
            $this->client,
            $request->getMethod(),
            $request->getRequestTarget(),
            $this->formatter->formatRequest($request)
        );

        $this->collector->addStack($stack);
        $eventName = sprintf('%s %s', $request->getMethod(), $request->getUri()->__toString());
        $event = $this->stopwatch->start($eventName, 'php_http.request');
        return $next($request)->then(function (ResponseInterface $response) use ($stack, $event) {
            $stack->setResponse($this->formatter->formatResponse($response));
            $stack->setResponseCode($response->getStatusCode());
            
            $event->stop();
            $stack->setDuration($event->getDuration());

            return $response;
        }, function (Exception $exception) use ($stack) {
            $stack->setResponse($this->formatter->formatException($exception));
            $stack->setFailed(true);

            $event->stop();
            $stack->setDuration($event->getDuration());

            throw $exception;
        });
    }
```

Such implementation does not require changes neither in symfony nor in the stopwatch plugin :)

**Comment by chr-hertel** [comment]
@fbourigault yes, would be an easy solution and more convincing than extending symfony's stopwatch. thanks :)
but IMO it's still a valid option to manipulate the response in stopwatch-plugin. we're just adding plugin related meta information to response.
@dbu @Nyholm let me know what your favorite choice is and i will adjust those PRs.

**Comment by dbu** [comment]
i think using the stopwatch directly instead of the plugin sounds better to me. much less complicated and achieves the same functionality in a clean way. at that point we would not need the stopwatch plugin for the bundle at all anymore, right?

**Comment by fbourigault** [comment]
The dependency will still be required. See https://github.com/php-http/HttplugBundle/blob/master/DependencyInjection/HttplugExtension.php#L185-L187.

**Comment by dbu** [comment]
oh good. then we don't need to decide if it would be a BC break to change the dependency...

**Comment by chr-hertel** [comment]
alright, great! :) i'll take care of this the next days

**Comment by Nyholm** [comment]
I've added this to the 1.5 milestone. Can you see why the tests are failing?

**Comment by fbourigault** [comment]
Could you also rebase your branch to get thinks working with `ProfileClient`?

**Comment by chr-hertel** [comment]
rebase is done and tests are not failing anymore. (honestly no clue what happened there.)

any ideas how to deal with the stopwatch-plugin? i think we need to disabled or even remove it because it interferes with the current implementation. but we could also stop using `debug.stopwatch` service in `ProfileClient` and instantiate a separate stopwatch.

**Comment by fbourigault** [comment]
About the stopwatch, I would remove the plugin and mesure time in the StackPlugin.

File: Collector/ProfileClient.php [hunk 2]
```diff
@@ -63,16 +72,20 @@ public function sendAsyncRequest(RequestInterface $request)
...
+            function (ResponseInterface $response) use ($event, $stack) {
+                $event->stop();
+                $this->collectResponseInformations($response, $event, $stack);
 
-            return $response;
-        }, function (\Exception $exception) use ($stack) {
-            $this->collectExceptionInformations($exception, $stack);
+                return $response;
+            }, function (\Exception $exception) use ($event, $stack) {
+                $this->collectExceptionInformations($exception, $event, $stack);
```
Comment (fbourigault) [inline Collector/ProfileClient.php]: Should we call `$event->stop()` here?

File: Collector/ProfileClient.php [hunk 3]
```diff
@@ -115,32 +131,48 @@ private function collectRequestInformations(RequestInterface $request, Stack $st
...
+    /**
+     * Generates the event name.
+     *
+     * @param RequestInterface $request
+     *
+     * @return string
+     */
+    private function getStopwatchEventName(RequestInterface $request)
+    {
+        return sprintf('%s %s', $request->getMethod(), $request->getUri()->__toString());
```
Comment (fbourigault) [inline Collector/ProfileClient.php]: The `-->toString()` is useless as `%s` will cast the URI as string.

**Comment by chr-hertel** [comment]
@fbourigault Thanks for your review! i wasn't able to reproduce your issues exactly but i had quite similar effects caused by duplicate stopwatch event names, so i improved `ProfileClient::getStopwatchEventName`.
can you please test, if that fixes your problem as well?

File: Collector/Collector.php [hunk 4]
```diff
@@ -108,4 +108,14 @@ public function getClientStacks($client)
             return $stack->getClient() == $client;
         });
     }
+
+    /**
+     * @return int
+     */
+    public function getDurationSum()
```
Comment (dbu) [inline Collector/Collector.php]: could we call this getTotalDuration?

**Comment by fbourigault** [comment]
Thank you @chr-hertel.

**Pull request merged**